### PR TITLE
Selectors: Add selectors for new sharepoint category

### LINKED ISSUE #2110: Selectors: Add Selectors for new SharePoint Category
Add selectors for SharePoint Category --> Pages

### LINKED ISSUE #2112: Selectors: Testing
- Related to #2110 
Add test into selectors package that ensures that  new selectors do not show up as `Unknown` within `path.Categories`

### PR DESCRIPTION
**Pull request by dadams39**
## Description
Expands SharePoint Selectors to include `SharePoint.Pages` within the Corso package.

## Does this PR need a docs update or release note?

- [x] :white_check_mark: Yes, it's included

## Type of change

-  [x] :sunflower: Feature
- [x] :bug: Bugfix

Prerequisites
-----
- [ ] Merge #2114 
- [x] Resolve #2112
## Issue(s)

* closes #2110<issue>
* related #2107

## Test Plan
Testing will 
- [x] :zap: Unit test

### CODE REVIEW

**Comment by dadams39** [comment]
Verified Corso command: `./corso backup create sharepoint --site "<siteID> "` still works as of [ea60b4e](https://github.com/alcionai/corso/pull/2111/commits/ea60b4e301e4183b72f48cb43388c60d63950d5d). On to adding tests.

**Comment by dadams39** [comment]
With current code  Corso binary fails with the following message:
```bash
Error: 1 error occurred:
	* Failed to run SharePoint backup for site <SITEID>: 2 errors occurred:
	* errors prevented the operation from processing
	* connecting to M365: building metadata path: unknown service/category combination "sharepoint"/"UnknownCategory
```

**Comment by dadams39** [comment]
This is not a complication from Incrementals. In the process of grabbing the manifest files,
able to short circuit prior to   `collectMetadata()` L372 
Screen Shot 2023-01-11 at 12 54 28 PM
and the operation is erroring when assigning the Pages to the SharePoint service.  
```bash
2 errors occurred:
	* errors prevented the operation from processing
	* backing up service data: getting subtree paths for bases: building path for service sharepoint category UnknownCategory: unknown service/category combination "sharepoint"/"UnknownCategory"
```




Screen Shot 2023-01-11 at 12.52.47 PM.png

**Comment by dadams39** [comment]
> This is not a complication from Incrementals. In the process of grabbing the manifest files, able to short circuit prior to `collectMetadata()` L372 Screen Shot 2023-01-11 at 12 54 28 PM and the operation is erroring when assigning the Pages to the SharePoint service.
> 
> ```shell
> 2 errors occurred:
> 	* errors prevented the operation from processing
> 	* backing up service data: getting subtree paths for bases: building path for service sharepoint category UnknownCategory: unknown service/category combination "sharepoint"/"UnknownCategory"
> ```
> 
> Screen Shot 2023-01-11 at 12.52.47 PM.png
- Issue #2112 to be added to the requirement for this PR to be merged.

**Comment by dadams39** [comment]
Error caused by misconfiguration of `LeafCat()`

File: src/pkg/selectors/sharepoint.go [hunk 1]
```diff
@@ -315,6 +353,8 @@ const (
 	SharePointListItem    sharePointCategory = "SharePointListItem"
 	SharePointLibrary     sharePointCategory = "SharePointLibrary"
 	SharePointLibraryItem sharePointCategory = "SharePointLibraryItem"
+	SharePointPage        sharePointCategory = "SharePointPage"
+	SharePointPageItem    sharePointCategory = "SharePointPageItem"
```
Comment (ryanfkeepers) [inline src/pkg/selectors/sharepoint.go:357]: We did a naming inversion with libraries in lists, because the "thing" itself (a List, or a Library) is a container.  Therefore we have, for example, Libraries and LibraryItems.

I think that pages, however, are more suited to the pattern we see elsewhere, like in exchange.  Most cases, the "thing" is the item.  Eg: Mail, or Contacts.  So we suffix "folder" (or whatever container) onto the thing.

Having a `PageFolder` that holds `Pages` as items seems to fit paradigm more closely.  Mostly because I'd expect someone who says, "i want to restore a page" to be referencing the file itself, and not the directory.

```suggestion
	SharePointPageFolder        sharePointCategory = "SharePointPageFolder"
	SharePointPage    sharePointCategory = "SharePointPage"
```

File: src/pkg/selectors/sharepoint.go [hunk 2]
```diff
@@ -291,6 +297,38 @@ func (s *sharePoint) LibraryItems(libraries, items []string, opts ...option) []S
...
+
+	return scopes
+}
+
+// PageItems produces one or more SharePoint page item scopes.
+// If any slice contains selectors.Any, that slice is reduced to [selectors.Any]
+// If any slice contains selectors.None, that slice is reduced to [selectors.None]
+// If any slice is empty, it defaults to [selectors.None]
+// options are only applied to the page scopes.
+func (s *sharePoint) PageItems(sites, pages []string, opts ...option) []SharePointScope {
```
Comment (ryanfkeepers) [inline src/pkg/selectors/sharepoint.go:320]:
```suggestion
func (s *sharePoint) PageItems(pages, items []string, opts ...option) []SharePointScope {
```

File: src/cli/utils/sharepoint_test.go [hunk 3]
```diff
@@ -17,6 +17,8 @@ func TestSharePointUtilsSuite(t *testing.T) {
 	suite.Run(t, new(SharePointUtilsSuite))
 }
 
+// Tests selector build for SharePoint properly
+// differentiates between the 3 categories: Pages, Libraries and Lists CLI
 func (suite *SharePointUtilsSuite) TestIncludeSharePointRestoreDataSelectors() {
```
Comment (ryanfkeepers) [inline src/cli/utils/sharepoint_test.go:22]: This test could use some cases that includes Pages.  (It could also use some cases for Lists, but that can be another PR).

File: src/pkg/selectors/selectors_test.go [hunk 4]
```diff
@@ -239,3 +240,126 @@ func (suite *SelectorSuite) TestSplitByResourceOnwer() {
 		})
 	}
 }
+
+// TestPathCategoriesAll includes all the included selectors do not return `Unknown` Category
```
Comment (ryanfkeepers) [inline src/pkg/selectors/selectors_test.go:244]:
```suggestion
// TestPathCategories verifies that no scope produces a `path.UnknownCategory`
```

File: src/pkg/selectors/selectors_test.go [hunk 5]
```diff
@@ -239,3 +240,126 @@ func (suite *SelectorSuite) TestSplitByResourceOnwer() {
...
+func (suite *SelectorSuite) TestPathCategoriesAll() {
```
Comment (ryanfkeepers) [inline src/pkg/selectors/selectors_test.go:245]:
```suggestion
func (suite *SelectorSuite) TestPathCategories_includes() {
```

File: src/pkg/selectors/selectors_test.go [hunk 6]
```diff
@@ -239,3 +240,126 @@ func (suite *SelectorSuite) TestSplitByResourceOnwer() {
...
+	users := []string{tester.LoadTestM365UserID(suite.T())}
```
Comment (ryanfkeepers) [inline src/pkg/selectors/selectors_test.go:246]: It's not necessary to load a user, since none of these values are checked against the tenant.  I'd suggest using a static string instead.

File: src/pkg/selectors/selectors_test.go [hunk 7]
```diff
@@ -239,3 +240,126 @@ func (suite *SelectorSuite) TestSplitByResourceOnwer() {
...
+				sel.Include(sel.EventCalendars([]string{"July"}, PrefixMatch()))
+				return &sel.Selector
+			},
+		},
+		{
+			name:  "SharePoint Pages",
+			isErr: assert.NoError,
+			getSelector: func(t *testing.T) *Selector {
+				sel := NewSharePointBackup(users)
+				sel.Include(sel.Pages([]string{"Something"}, SuffixMatch()))
```
Comment (ryanfkeepers) [inline src/pkg/selectors/selectors_test.go:319]: Any particular reason to use suffix match here?

File: src/pkg/selectors/selectors_test.go [hunk 8]
```diff
@@ -239,3 +240,126 @@ func (suite *SelectorSuite) TestSplitByResourceOnwer() {
...
+			getSelector: func(t *testing.T) *Selector {
+				return &Selector{}
+			},
+		},
+		{
+			name:  "Mail_B",
+			isErr: assert.NoError,
+			getSelector: func(t *testing.T) *Selector {
+				sel := NewExchangeBackup(users)
+				sel.Include(sel.MailFolders([]string{"MailFolder"}, PrefixMatch()))
```
Comment (ryanfkeepers) [inline src/pkg/selectors/selectors_test.go:265]: I'd suggest expanding this test to include a specific item in addition to the folder type.  There should probably be a test for filters as well, but this is enough coverage for now.
```suggestion
				sel.Include(
				    sel.MailFolders([]string{"MailFolder"}, PrefixMatch()),
				    sel.Mail([]string{"MailFolder2"}, []string{"Mail"}),
				)
```

**Review by ryanfkeepers — APPROVED** [review]

File: src/pkg/selectors/selectors_test.go (hunk 7 above)
Comment (dadams39) [reply src/pkg/selectors/selectors_test.go:319]: None

File: src/pkg/selectors/selectors_test.go (hunk 8 above)
Comment (dadams39) [reply src/pkg/selectors/selectors_test.go:265]: There are specific filter tests to ensure that abstraction was working. The gotcha that I found while looking into #2112 was that a `sel.Include(<topology>` was returning a category, `unknown` which causes an operation to fail. Will include a few others as random testing.

File: src/pkg/selectors/sharepoint.go (hunk 2 above)
Comment (dadams39) [reply src/pkg/selectors/sharepoint.go]: Resolved within [82b5d21](https://github.com/alcionai/corso/pull/2111/commits/82b5d21f190583a634c1fc0fabeadcf50aabbe11)

File: src/pkg/selectors/sharepoint.go (hunk 1 above)
Comment (dadams39) [reply src/pkg/selectors/sharepoint.go]: Resolved within [82b5d21](https://github.com/alcionai/corso/pull/2111/commits/82b5d21f190583a634c1fc0fabeadcf50aabbe11)

File: src/pkg/selectors/selectors_test.go (hunk 8 above)
Comment (dadams39) [reply src/pkg/selectors/selectors_test.go:264]: Resolved within [fa9b7ad](https://github.com/alcionai/corso/pull/2111/commits/fa9b7ad72227a7b60a2fe1532d5112eb3a430654)

File: src/pkg/selectors/selectors_test.go (hunk 6 above)
Comment (dadams39) [reply src/pkg/selectors/selectors_test.go]: Resolved within [fa9b7ad](https://github.com/alcionai/corso/pull/2111/commits/fa9b7ad72227a7b60a2fe1532d5112eb3a430654)

File: src/pkg/selectors/selectors_test.go (hunk 5 above)
Comment (dadams39) [reply src/pkg/selectors/selectors_test.go]: Resolved within [fa9b7ad](https://github.com/alcionai/corso/pull/2111/commits/fa9b7ad72227a7b60a2fe1532d5112eb3a430654)

File: src/pkg/selectors/selectors_test.go (hunk 4 above)
Comment (dadams39) [reply src/pkg/selectors/selectors_test.go]: Resolved within [fa9b7ad](https://github.com/alcionai/corso/pull/2111/commits/fa9b7ad72227a7b60a2fe1532d5112eb3a430654)

File: src/cli/utils/sharepoint_test.go (hunk 3 above)
Comment (dadams39) [reply src/cli/utils/sharepoint_test.go:22]: There are some scattered Lists tests within

Comment (dadams39) [reply src/cli/utils/sharepoint_test.go:22]: Resolved within [e441252](https://github.com/alcionai/corso/pull/2111/commits/e441252f06d2b2cbc588f09d081135177e949034)
